Stratis.Bitcoin/Features/Consensus/TimeSyncBehavior.cs

```diff
@@ -0,0 +1,458 @@
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly HashSet<IPAddress> inboundSampleSources;
+
+        /// <summary>List of IP addresses of peers that provided samples in <see cref="outboundTimestampOffsets"/>.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly HashSet<IPAddress> outboundSampleSources;
+
+        /// <summary><c>true</c> if the warning loop has been started, <c>false</c> otherwise.</summary>
+        public bool WarningLoopStarted { get; private set; }
+
+        /// <summary><c>true</c> if the time sync with peers has been switched off, <c>false</c> otherwise.</summary>
+        public bool SwitchedOff { get; private set; }
+
+        /// <summary>
+        /// <c>true</c> if the reason for switching the time sync feature off was that <see cref="timeOffset"/> 
+        /// went over the maximal allowed value, <c>false</c> otherwise.
+        /// </summary>
+        public bool SwitchedOffLimitReached { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="nodeLifetime">Global application life cycle control - triggers when application shuts down.</param>
+        /// <param name="asyncLoopFactory">Factory for creating background async loop tasks.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public TimeSyncBehaviorState(IDateTimeProvider dateTimeProvider, INodeLifetime nodeLifetime, IAsyncLoopFactory asyncLoopFactory, ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger(GetType().FullName);
+            this.dateTimeProvider = dateTimeProvider;
+            this.nodeLifetime = nodeLifetime;
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.inboundTimestampOffsets = new CircularArray<TimestampOffsetSample>(MaxInboundSamples);
+            this.outboundTimestampOffsets = new CircularArray<TimestampOffsetSample>(MaxOutboundSamples);
+            this.inboundSampleSources = new HashSet<IPAddress>();
+            this.outboundSampleSources = new HashSet<IPAddress>();
+
+            this.timeOffset = TimeSpan.Zero;
+            this.WarningLoopStarted = false;
+            this.SwitchedOff = false;
+            this.SwitchedOffLimitReached = false;
+        }
+
+        /// <inheritdoc />
+        public bool AddTimeData(IPAddress peerAddress, TimeSpan offsetSample, bool isInboundConnection)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:{3},{4}:{5})", nameof(peerAddress), peerAddress, nameof(offsetSample), offsetSample.TotalSeconds, nameof(isInboundConnection), isInboundConnection);
+
+            bool res = false;
+
+            bool startWarningLoopNow = false;
```

Comment:
This is a classic double lock scenario
 ```
if (!this.SwitchedOff)
{
      lock (this.lockObject)
      {
             if (!this.SwitchedOff)
             {
```

Reply:
tx, https://github.com/stratisproject/StratisBitcoinFullNode/issues/461